src/AdvancedVI.jl

```diff
@@ -1,270 +1,147 @@
+This function needs to be implemented only if `obj` is stateful.
+
+!!! warning
+    This is an internal function. Thus, the signature is subject to change without
+    notice.
 """
-function vi end
-
-function update end
-
-# default implementations
-function grad!(
-    vo,
-    alg::VariationalInference{<:ForwardDiffAD},
-    q,
-    model,
-    θ::AbstractVector{<:Real},
-    out::DiffResults.MutableDiffResult,
-    args...
-)
-    f(θ_) = if (q isa Distribution)
-        - vo(alg, update(q, θ_), model, args...)
-    else
-        - vo(alg, q(θ_), model, args...)
-    end
+init(
+    rng::Random.AbstractRNG,
+    obj::AbstractVariationalObjective,
+    λ::AbstractVector,
+    restructure
+) = nothing
 
-    # Set chunk size and do ForwardMode.
-    chunk_size = getchunksize(typeof(alg))
-    config = if chunk_size == 0
-        ForwardDiff.GradientConfig(f, θ)
-    else
-        ForwardDiff.GradientConfig(f, θ, ForwardDiff.Chunk(length(θ), chunk_size))
-    end
-    ForwardDiff.gradient!(out, f, θ, config)
-end
+"""
+    estimate_gradient!(
+        rng         ::Random.AbstractRNG,
+        prob,
+        adbackend   ::ADTypes.AbstractADType,
+        obj         ::AbstractVariationalObjective,
+        obj_state,
+        λ           ::AbstractVector,
+        restructure,
+        out         ::DiffResults.MutableDiffResult
+    )
+
+Estimate (possibly stochastic) gradients of the objective `obj` targeting `prob` with respect to the variational parameters `λ` using the automatic differentiation backend `adbackend`.
+The estimated objective value and gradient are then stored in `out`.
+If the objective is stateful, `obj_state` is its previous state, otherwise, it is `nothing`.
+
+# Returns
+- `out`: The `MutableDiffResult` containing the objective value and gradient estimates.
+- `obj_state`: The updated state of the objective estimator.
+- `stat`: Statistics and logs generated during estimation. (Type: `<: NamedTuple`)
+
+!!! warning
+    This is an internal function. Thus, the signature is subject to change without
+    notice.
+"""
+function estimate_gradient! end
 
-function grad!(
-    vo,
-    alg::VariationalInference{<:TrackerAD},
-    q,
-    model,
-    θ::AbstractVector{<:Real},
-    out::DiffResults.MutableDiffResult,
-    args...
-)
-    θ_tracked = Tracker.param(θ)
-    y = if (q isa Distribution)
-        - vo(alg, update(q, θ_tracked), model, args...)
-    else
-        - vo(alg, q(θ_tracked), model, args...)
-    end
-    Tracker.back!(y, 1.0)
+# ADVI-specific interfaces
+abstract type AbstractEntropyEstimator end
 
-    DiffResults.value!(out, Tracker.data(y))
-    DiffResults.gradient!(out, Tracker.grad(θ_tracked))
-end
+# entropy.jl must preceed advi.jl
+include("objectives/elbo/entropy.jl")
+include("objectives/elbo/advi.jl")
 
+export
+    ADVI,
+    ClosedFormEntropy,
+    StickingTheLandingEntropy,
+    MonteCarloEntropy
```

Comment:
I believe convention is to have the exports before the files rather than after (it's not well-established, but personally find it a bit nicer that you can see which public-facing methods are being defined immediately rather having looked at all the includes)
```suggestion
export
    ADVI,
    ClosedFormEntropy,
    StickingTheLandingEntropy,
    MonteCarloEntropy
# entropy.jl must preceed advi.jl
include("objectives/elbo/entropy.jl")
include("objectives/elbo/advi.jl")
```

Reply:
I see, I didn't know about that. Thanks!